cijoe/scripts/fio_latency.py

```diff
@@ -0,0 +1,404 @@
+from enum import Enum
+from itertools import product
+from pathlib import Path
+from typing import Any, Dict, Iterator, List, Optional
+
+from cijoe.core.command import Cijoe
+
+LATENCY_TEST_SIZE: str = "1G"
+LATENCY_TEST_ROOT_DIR: Path = Path("/tmp/")
+
+CijoeEngines = Dict[str, Dict[str, str]]
+CijoeDevices = List[Dict[str, str | int | List[str]]]
+
+
+@dataclass
+class Device:
+    key: str
+    nsid: int
+    uri: str
+    labels: List[str]
+    driver_attachment: str
+
+
+def determine_devices(devices: CijoeDevices) -> Iterator[Device]:
+    for device in devices:
+        assert "key" in device and isinstance(device["key"], str)
+        assert "nsid" in device and isinstance(device["nsid"], int)
+        assert "uri" in device and isinstance(device["uri"], str)
+        assert "labels" in device and isinstance(device["labels"], list)
+        assert "driver_attachment" in device and isinstance(
+            device["driver_attachment"], str
+        )
+        yield Device(
+            key=device["key"],
+            nsid=device["nsid"],
+            uri=device["uri"],
+            labels=device["labels"],
+            driver_attachment=device["driver_attachment"],
+        )
+
+
+class Engines(Enum):
+    NULL = "null"
+    IO_URING = "io_uring"
+    IO_URING_CMD = "io_uring_cmd"
+    LIBAIO = "libaio"
+    POSIXAIO = "posixaio"
+    NIL = "nil"
+    SPDK_NVME = "spdk_nvme"
+    SPDK_BDEV = "spdk_bdev"
+    XNVME_IO_URING = "xnvme_io_uring"
+    XNVME_IO_URING_CMD = "xnvme_io_uring_cmd"
+    XNVME_LIBAIO = "xnvme_libaio"
+    XNVME_POSIXAIO = "xnvme_posixaio"
+    XNVME_NULL = "xnvme_null"
+    XNVME_SPDK = "xnvme_spdk"
+
+
+@dataclass
+class Engine:
+    name: str
+    device: Device
+    cijoe: Cijoe
+
+    @property
+    def name_id(self) -> str:
+        return f"{self.name}"
+
+    @property
+    def ioengine(self) -> str:
+        return self.name
+
+    def env(self) -> Dict[str, str]:
+        """Environment variables used in execution of test."""
+        return {}
+
+    @property
+    def filename(self) -> str:
+        return self.device.uri
+
+    @property
+    def direct(self) -> str:
+        """Value for the `--direct` fio flag."""
+        return "1"
+
+    def binary(self, bin: str) -> str:
+        """Path to fio binary"""
+        return bin
+
+    def extra_args(self) -> Dict[str, str]:
+        """Method that can be overridden to add extra arguments when running fio.
+        e.g. xnvme backend engine when it applies.
+        """
+        return {}
+
+    def prepare(self) -> None:
+        """Method that will run in preparation for execution fio test.
+        e.g. is used to setup nvme devices.
+        """
+
+
+class KernelEngine(Engine):
+    def prepare(self) -> None:
+        self.cijoe.run(" ".join(["xnvme-driver", "reset"]))
+
+
+class UserSpaceEngine(Engine):
+    def prepare(self) -> None:
+        self.cijoe.run(" ".join(["xnvme-driver"]))
+
+
+@dataclass
+class BuiltinEngine(KernelEngine):
+    ...
```

Comment:
Perhaps this can be removed - and anything pointing to it can point to `KernelEngine` instead?

Reply:
Let do it. It seems superfluous now.